docs/proposals/predictive-horizontal-pod-autoscaling.md

```diff
@@ -0,0 +1,232 @@
+meaning there was 10 minutes in which pods were operating outside of target
+resource consumption). Predictive horizontal pod autoscaling can
+decrease the impact of pod initialization times on autoscaling responsiveness and thus
+decrease the amount of time pods spend operating outside of their target
+resource consumption.
+
+## Predictive Autoscaling Algorithm
+
+Currently, horizontal pod autoscaling is implemented as a control loop.
+Every period, the autoscaler checks all machines specified by the
+`Status.PodSelector` of the Scale subresource, and checks their metrics through
+the [Compute Resource Metrics Api](compute-resource-metrics-api.md).
+It computes the `TargetNumOfPods` based on the previously described algorithm, with the following:
+
+```
+TargetNumOfPods = ceil(sum(CurrentPodsResourceConsumption) / TargetResourceConsumption)
+```
+
+Predictive horizontal autoscaling will use a similar algorithm with one
+significant difference. Rather than calculating based on the
+`CurrentPodsResourceConsumption`, we will calculate based on
+`FuturePodsResourceConsumption`, which we initially define as
+`FuturePodsResourceConsumption = CurrentPodResourceConsumption +
+(PodsAverageInitializationTime * DerivativeCurrentPodsResourceConsumptionPerSecond)`.
+The calculation for `FuturePodsResourceConsumption` could become more complex,
+incorporating increasingly complex methods of prediction, in the future.
+We now calculate `TargetNumOfPods` with the following:
+
+```
+TargetNumOfPods = ceil(sum(FuturePodsResourceConsumption)) / Target
+```
+
+The modifications for this option will predominantly occur in
+`pkg/controllers/podautoscaler/horizontal.go` in the
+`computeReplicasForCPUUtilization` function. The new algorithm will be used
+only if predictive auto-scaling is turned on (as discussed in the next
+section).
+
+## Implementation
+
+### Turning on Prediction
+
+Initially, the configuration value for enabling predictive autoscaling will be recorded in
+the `HorizontalPodAutoscaler.ObjectMeta.Annotations`.
+
+This annotation can be done via the following command:
+
+```sh
+kubectl annotate hpa foo predictive=’true’
+```
+
+### Other modifications
+
+Additionally, the algorithm for predictive autoscaling requires two new pieces
+of information: a pod’s initialization time and the rate of change for the
+metric upon which scaling behavior is determined (currently only CPU). These
+two values are currently untracked; the following two sections describe the
+changes to record them.
+
+#### Recording `AveragePodInitializationTime`
+
+`AveragePodInitializeTime` has a couple of important nuances. First, it is not
+simply the amount of time between when a pod is ordered to be created and when
+it is created. Rather, we want to record the amount of time from when a pod is
+ordered to be created and when it is ready to perform the task for which it is
+created. For example, a pod may be created, but if the purpose of the pod is to
```

Comment:
Sounds perfect - I had been thinking the readiness probe would be a good fit, so its good to have that confirmed!